### PR TITLE
PR to master - Introduce Recommended access order (RAO) via EA

### PR DESCRIPTION
**Pull request by y-sasn**
# Summary of changes

This PR is related to #287 .
Please refer to this comment: https://github.com/LinearTapeFileSystem/ltfs/pull/287#issuecomment-848525367

Introduce Recommended access order (RAO) support via EA by `ltfs.vendor.IBM.rao`.

# Description

This update introduce the support of the RAO (GRAO, RRAO) command, which was introduced from LTO-9.
Only linux scsi generic (sg) driver is supported in this fix.

## How to use

1. Set a file path to a correct GRAO Parameter Data via VEA on mount point.
2. A `xxx.out` file containing the RAO List will be made in the same directory.

example:
```
attr -s ltfs.vendor.IBM.rao -V ~/xxx <mount dir>
# The result will be created in `~/xxx.out`
```

## Type of change

- New feature (non-breaking change which adds functionality)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have confirmed my fix is effective or that my feature works

### CODE REVIEW

File: src/libltfs/tape.c [hunk 1]
```diff
@@ -3570,3 +3570,48 @@ int tape_set_profiler(struct device_data *dev, char *work_dir, bool enable)
...
+		return EDEV_INTERNAL_ERROR;
+	}
+
+	/* run GRAO */
+	ret = dev->backend->grao(dev->backend_data, (const unsigned char *)rao->in_buf, rao->num_of_files);
+	if (ret < 0) {
+		ltfsmsg(LTFS_ERR, 17278E, "GRAO", ret); //GRAO command returns error
+		return ret;
+	}
+	else if (rao->num_of_files <= 0) {
```
Comment (piste-jp-ibm) [inline src/libltfs/tape.c:3603]: Please do not make a line break between `}` and `else`.

File: src/libltfs/ltfs.h [hunk 2]
```diff
@@ -698,6 +707,11 @@ int ltfs_print_device_list(struct tape_ops *ops);
 void ltfs_enable_livelink_mode(struct ltfs_volume *vol);
 
 int ltfs_profiler_set(uint64_t source, struct ltfs_volume *vol);
+
+int ltfs_get_rao_list(char *path, struct ltfs_volume *vol);
+int _ltfs_write_rao_file(char *write_data, char *file_path, size_t *write_size);
+int _ltfs_read_rao_file(char *read_data, uint32_t *num_of_files, char *file_path);
```
Comment (piste-jp-ibm) [inline src/libltfs/ltfs.h:713]: Same as `_ltfs_write_rao_file()`

**Review by piste-jp-ibm — CHANGES_REQUESTED** [review]
Could you check my comments?

They are not major but not minor...

File: src/tape_drivers/linux/sg/sg_tape.c [hunk 3]
```diff
@@ -1169,6 +1169,80 @@ static int _register_key(void *device, unsigned char *key)
...
+	int ret_ep = DEVICE_GOOD;
+
+	struct sg_data *priv = (struct sg_data*)device;
+	sg_io_hdr_t req;
+	unsigned char sense[MAXSENSE];
+	unsigned char cdb[CDB12_LEN];
+	int timeout;
+	char cmd_desc[COMMAND_DESCRIPTION_LENGTH] = "RRAO";
+	char *msg = NULL;
+
```
[the remark this answers is not in the document]
Comment (piste-jp-ibm) [reply src/tape_drivers/linux/sg/sg_tape.c:1191]: I think there is no need to include to this PR.

We can handle this thing as an enhancement later.

File: src/libltfs/tape.c (hunk 1 above)
Comment (y-sasn) [reply src/libltfs/tape.c]: modified in https://github.com/LinearTapeFileSystem/ltfs/pull/288/commits/87b5a2d44974149001f8f29ae12a3718b102e69f

File: src/libltfs/ltfs.h [hunk 4]
```diff
@@ -698,6 +707,11 @@ int ltfs_print_device_list(struct tape_ops *ops);
 void ltfs_enable_livelink_mode(struct ltfs_volume *vol);
 
 int ltfs_profiler_set(uint64_t source, struct ltfs_volume *vol);
+
+int ltfs_get_rao_list(char *path, struct ltfs_volume *vol);
+int _ltfs_write_rao_file(char *write_data, char *file_path, size_t *write_size);
```
[the remark this answers is not in the document]
Comment (y-sasn) [reply src/libltfs/ltfs.h]: modified in https://github.com/LinearTapeFileSystem/ltfs/pull/288/commits/87b5a2d44974149001f8f29ae12a3718b102e69f

File: src/libltfs/ltfs.h (hunk 2 above)
Comment (y-sasn) [reply src/libltfs/ltfs.h]: modified in https://github.com/LinearTapeFileSystem/ltfs/pull/288/commits/87b5a2d44974149001f8f29ae12a3718b102e69f

File: src/tape_drivers/linux/sg/sg_tape.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (y-sasn) [reply src/tape_drivers/linux/sg/sg_tape.c:1191]: The RRAO command itself will fail for unsupported drives, causing all RAO command to be rejected.
I'll leave the code as-is for this PR, but yes, I agree that adding device check here is most definitely preferred.

**Review by piste-jp-ibm — APPROVED** [review]
It looks good to me.

**Pull request merged**